### DISCUSSION TITLE
TypeError: r is not a function

### DISCUSSION THREAD

**Discussion by EncryptMan** [nextauthjs/next-auth#9251] [Help]
I am facing many problems with next-auth@beta and next js 14. After redirect back from discord oauth I am getting this error
"TypeError: r is not a function"

<img width="1221" alt="Screenshot 2023-11-27 at 10 47 13" src="https://github.com/nextauthjs/next-auth/assets/136573652/10273872-27d5-44f2-b82d-3e4f380f1ff1">
<img width="1040" alt="Screenshot 2023-11-27 at 10 47 03" src="https://github.com/nextauthjs/next-auth/assets/136573652/89228570-f4a5-4ec5-a6c4-088e3ffffc74">

**Comment by kaushikc44** [comment]
If your using App Router I recommend to watch this video https://youtu.be/md65iBX5Gxg?si=Qa2vL0uxL4cYj8pt and also you don't need to use the beta version they are unstable I figured it out yesterday. So saving you the trouble

**Reply by EncryptMan** [reply]
Which version of next-auth will work with nextjs 14?

**Comment by aronianm** [comment]
I had "next-auth": "^5.0.0-beta.4",
and downgrading to "next-auth": "4.24.5" that was released 2 months ago fixed this issue for me

**Reply by cdlftw** [reply]
Fixed the issue for me too.  Thanks!

**Reply by TheOrcDev** [reply]
Yep! This one fixed the problem! Thanks @aronianm !!

**Reply by jylee311** [reply]
Thank you very much !!!!!

**Reply by tjulioh** [reply]
Thanks, fixed the issue for me.

**Reply by Fritze1602** [reply]
same here. Thanks a million for this. Fixed a bunch of strange things.

**Reply by Nemidas7734** [reply]
Fixed issue for me , thanks a lot

**Reply by aronianm** [reply]
@TheOrcDev @jylee311 @tjulioh @Fritze1602  @Nemidas7734 

https://github.com/nextauthjs/next-auth/discussions/9251#discussioncomment-8399500

This should be your approach instead of downgrading.

**Reply by OlgaS-codes** [reply]
Thanks! Fixed the issue!

**Reply by 0nexo0** [reply]
Thank You So Much Bro Really Helpfull ❤️

**Comment by avelzo** [comment]
uninstalling the beta version("^5.0.0-beta.4") and reinstalling stable version("4.24.5") worked for me too.
Thank you guys

**Comment by TheOrcDev** [comment]
> 4.24.5
Here is the file `/api/auth/[...nextauth]/route.ts` that is working for me in this version:

```
import NextAuth from "next-auth";
import GoogleProvider from "next-auth/providers/google";

export const authOptions = {
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
  ],
};
const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };
```

After this just hit `api/auth/signin` URL

**Reply by vaibhavprasad54** [reply]
This worked for me as well, in case anyone needs to know I am using Next.js v14.1.4 and downgraded Next-auth to v4.24.5 as @aronianm mentioned.
Also check your layout.js/ts file, make sure you've correctly imported SessionProvider as a wrapper and not directly because layout is a server component.

**Reply by ness-e** [reply]
You are the best bro, I have been looking for comments close to the date for a week and this is the first one I have found that helps me, I am starting and I have many errors

**Comment by BrofessorNFT** [comment]
This helped me too! So what is wrong with the beta version of next-auth. This versions hell with nextjs-14 drives me crazy

**Reply by EncryptMan** [reply]
I don't know. I initially used next-auth@beta because official Learn Nextjs guide stated that only beta version is compatible with Next js 14.

**Comment by Zelzazor** [comment]
Since you are using next auth v5, I've solved it like this: 

- Change your `auth.ts` file to include the GET and POST methods:
```ts
export const { auth, signIn, signOut, handlers: { GET, POST } } = NextAuth({
  // Your Auth config here
});
```
- Now, on your `route.ts` file, add this single line:

```ts
export { GET, POST } from "@/auth"; // change the route to the auth.ts path if it's not the same
```

You have to make sure that all of your config for auth is inside `auth.ts` and nowhere else.
Auth.js pretty much overhauled its entire library on this `auth` method, replacing most of the options for server side handlers. Check this guide for more details: [https://authjs.dev/guides/upgrade-to-v5](https://authjs.dev/guides/upgrade-to-v5)

**Reply by Vegas97** [reply]
Man you saved my day tnx ! 
AND this works with beta !
{
    "next": "14.1.3",
    "next-auth": "^5.0.0-beta.15",
    "react": "^18",
    "react-dom": "^18"
}

**Reply by ishwarrimal** [reply]
Man... I spent the last 2 days debugging this.
New to nextjs and the tutorial is very misleading.
This just saved my life. (yeah crying right now)

**Reply by stabla** [reply]
This.

The documentation from NextAuth/NextJS are so terrible... lost 2 days on this topic... thank you, you saved me a lot of pain!

**Reply by parthematics** [reply]
@Zelzazor this worked for me! Thank you so much, wasted too much time on this...

**Reply by vladimirmazarakis** [reply]
> Since you are using next auth v5, I've solved it like this:
> 
> * Change your `auth.ts` file to include the GET and POST methods:
> 
> ```ts
> export const { auth, signIn, signOut, handlers: { GET, POST } } = NextAuth({
>   // Your Auth config here
> });
> ```
> 
> * Now, on your `route.ts` file, add this single line:
> 
> ```ts
> export { GET, POST } from "@/auth"; // change the route to the auth.ts path if it's not the same
> ```
> 
> You have to make sure that all of your config for auth is inside `auth.ts` and nowhere else. Auth.js pretty much overhauled its entire library on this `auth` method, replacing most of the options for server side handlers. Check this guide for more details: https://authjs.dev/guides/upgrade-to-v5

It makes sense, for all the other guys blaming the beta be a little bit more careful with your code :)

**Reply by HansiDerSweater** [reply]
Thank you Sir, this helped me! v5.0.0 beta 16

**Reply by nihal-dhore** [reply]
Thanks man that worked. I think that's why the say read docs carefully!

**Reply by nusratnova5** [reply]
Thank you

**Comment by joaopedropastura** [comment]
try this link: https://authjs.dev/getting-started/migrating-to-v5

i got the same error and its work for me.

problably u forget auth.ts file main folder project

**Reply by himanshu-tyd** [reply]
Thanks brother this work for me

just by installing "^5.0.0-beta.20" docs " check this link -> https://authjs.dev/getting-started/migrating-to-v5#authenticating-server-side

**Comment by lobotomoe** [comment]
In my case reason of this error is replacement from

```ts
const {
  handlers: { GET, POST },
  auth,
  signIn,
  signOut,
} = NextAuth(authConfig); // static object
```

to

```ts

const buildAuthConfig = (request?: NextRequest) => { // my logic }
const {
  handlers: { GET, POST },
  auth,
  signIn,
  signOut,
} = NextAuth(buildAuthConfig);
```

And I don't know how to fix that, return back to static config for now.

**Reply by kelvinpraises** [reply]
https://github.com/nextauthjs/next-auth/issues/11544

**Comment by eunjin0212** [comment]
I resolve this issue!
I got the same error and it works for me.

```typescript
// src/auth/index.ts
import NextAuth, { NextAuthConfig } from "next-auth";
import Google from 'next-auth/providers/google';

export const BASE_PATH = "/api/auth";
const authOptions: NextAuthConfig = {
  providers: [
    Google({
      clientId: process.env.GOOGLE_CLIENT_ID ?? '',
      clientSecret: process.env.GOOGLE_CLIENT_SECRET ?? ''
    })
  ],
  basePath: BASE_PATH,
};

export const { handlers: { GET, POST }, auth, signIn, signOut } = NextAuth(authOptions);
```

```typescript
// src/auth/helpers.ts
"use server";

import { signIn as SignIn, signOut as SignOut } from "./index";

export async function signIn() {
  await SignIn();
}

export async function signOut() {
  await SignOut();
}
```

If you want to see whole code, you can visit my [repository commit log](https://github.com/eunjin0212/side-next-todo/commit/537b0870b4c25c6c4ff752cf1291b79ada264dd0)

**Comment by CollinsKippy** [comment]
Mine occurred due to the AI Coding assistant (Claude - Haiku) adding a 'use client' directive to the route.ts handler file. Check for this in your route handlers. (I am using "next": "14.2.18", with "next-auth": "^5.0.0-beta.25").
